panther_hardware_interfaces/include/panther_hardware_interfaces/gpio_driver.hpp

```diff
@@ -1,33 +1,101 @@
 #ifndef PANTHER_HARDWARE_INTERFACES__GPIO_CONTROLLER_HPP_
 #define PANTHER_HARDWARE_INTERFACES__GPIO_CONTROLLER_HPP_
 
+#include <atomic>
 #include <thread>
+#include <vector>
+#include <poll.h>
 
+#include <boost/function.hpp>
 #include <gpiod.hpp>
 
 namespace panther_hardware_interfaces
 {
+enum class GPIOpins {
+  WATCHDOG = 0,
+  E_STOP_RESET,
+  MOTOR_DRIVER_EN,
+  VMOT_ON,
+  FAN_SW,
+  LED,
+  TEST_IN,
+  UNKNOWN
+};
+
+GPIOpins operator+(const GPIOpins & pin, int int_val);
 
-// Placeholder GPIO class used for simple enabling, will be replaced by proper implementation in the future
+struct GPIOinfo
+{
+  std::string name;
+  gpiod::line::direction direction;
+  gpiod::line::bias bias = gpiod::line::bias::AS_IS;
+  gpiod::line::value init_value = gpiod::line::value::INACTIVE;
+  gpiod::line::value value = gpiod::line::value::INACTIVE;
+  gpiod::line::offset offset = -1;
+};
 
-enum class GPIOPins { VMOT_ON = 6, MOTOR_DRIVER_EN = 23, WATCHDOG = 14, E_STOP_RESET = 27 };
+struct OperationResult
+{
+  bool success;
+  std::string message = "";
+};
 
 class GPIOController
 {
 public:
   GPIOController();
   ~GPIOController();
 
-  void Start();
+  bool start() { return watchdog_on(); };
+  bool motors_enable(bool enable) { return set_pin_value(GPIOpins::VMOT_ON, enable); }
+  bool fan_enable(bool enable) { return set_pin_value(GPIOpins::FAN_SW, enable); }
+  bool led_enable(bool enable) { return set_pin_value(GPIOpins::LED, enable); }
+  OperationResult e_stop_trigger();
+  OperationResult e_stop_reset();
+
+  boost::function<void(const GPIOinfo & gpio_info)> publish_gpio_state_callback;
 
 private:
-  std::unique_ptr<gpiod::chip> chip_;
-  std::unique_ptr<gpiod::line> vmot_;
-  std::unique_ptr<gpiod::line> motor_driver_;
-  std::unique_ptr<gpiod::line> e_stop_reset_;
-  std::unique_ptr<gpiod::line> watchdog_;
+  std::unique_ptr<gpiod::line_request> create_line_request(gpiod::chip & chip, GPIOpins pins);
+  std::unique_ptr<gpiod::line_request> create_line_request(
+    gpiod::chip & chip, const std::vector<GPIOpins> & pins);
+
+  void configure_line_request(gpiod::chip & chip, gpiod::request_builder & builder, GPIOpins pin);
+  gpiod::line_settings generate_line_settings(const GPIOinfo & pin_info);
+  GPIOpins get_pin_from_offset(gpiod::line::offset offset) const;
+
+  void change_control_pin_direction(GPIOpins pin, gpiod::line::direction direction);
+  bool get_pin_value(GPIOpins pin);
+  bool set_pin_value(GPIOpins pin, bool value);
+  bool watchdog_on();
+  bool watchdog_off();
+  void watchdog_thread();
+  bool is_watchdog_thread_running() const;
+  void gpio_monitor_on();
+  void gpio_monitor_off();
+  void monitor_async_events();
+  void handle_edge_event(const gpiod::edge_event & event);
 
+  std::map<GPIOpins, GPIOinfo> gpio_info_{
```

Comment:
It would be better not to use map - we are targeting RT, and it isn't really safe to use maps. We can use vector instead.